File: C++/11 - Graphs - Edge List, Adjacency Matrix, Adjacency List, DFS, BFS - Greg Hogg DSA Course Materials Lecture 11.cpp
// Graphs - Edge List, Adjacency Matrix, Adjacency List, DFS, BFS - Greg Hogg DSA Course Materials Lecture 11

#include <iostream>
#include <vector>
#include <queue>
#include <stack>
#include <unordered_map>
#include <unordered_set>

// Build Adjacency Matrix
std::vector<std::vector<int>> buildAdjacencyMatrix(int n, const std::vector<std::pair<int, int>>& edges) {
    std::vector<std::vector<int>> matrix(n, std::vector<int>(n, 0));
    for (const auto& edge : edges) {
        matrix[edge.first][edge.second] = 1;
    }
    return matrix;
}

// Build Adjacency List
std::unordered_map<int, std::vector<int>> buildAdjacencyList(const std::vector<std::pair<int, int>>& edges) {
    std::unordered_map<int, std::vector<int>> adjList;
    for (const auto& edge : edges) {
        adjList[edge.first].push_back(edge.second);
    }
    return adjList;
}

// Recursive DFS
void dfsRecursive(int node, const std::unordered_map<int, std::vector<int>>& adjList, std::unordered_set<int>& seen) {
    if (seen.count(node)) return;
    seen.insert(node);
    std::cout << node << " ";
    if (adjList.find(node) != adjList.end()) {
        for (int neighbor : adjList.at(node)) {
            dfsRecursive(neighbor, adjList, seen);
        }
    }
}

// Iterative DFS
void dfsIterative(int startNode, const std::unordered_map<int, std::vector<int>>& adjList) {
    std::unordered_set<int> seen;
    std::stack<int> stack;
    stack.push(startNode);
    seen.insert(startNode);

    while (!stack.empty()) {
        int node = stack.top();
        stack.pop();
        std::cout << node << " ";
        if (adjList.find(node) != adjList.end()) {
            for (int neighbor : adjList.at(node)) {
                if (!seen.count(neighbor)) {
                    seen.insert(neighbor);
                    stack.push(neighbor);
                }
            }
        }
    }
}

// BFS
void bfs(int startNode, const std::unordered_map<int, std::vector<int>>& adjList) {
    std::unordered_set<int> seen;
    std::queue<int> queue;
    queue.push(startNode);
    seen.insert(startNode);

    while (!queue.empty()) {
        int node = queue.front();
        queue.pop();
        std::cout << node << " ";
        if (adjList.find(node) != adjList.end()) {
            for (int neighbor : adjList.at(node)) {
                if (!seen.count(neighbor)) {
                    seen.insert(neighbor);
                    queue.push(neighbor);
                }
            }
        }
    }
}

// Print Matrix
void printMatrix(const std::vector<std::vector<int>>& matrix) {
    for (const auto& row : matrix) {
        for (int val : row) {
            std::cout << val << " ";
        }
        std::cout << std::endl;
    }
}

// Print Adjacency List
void printAdjacencyList(const std::unordered_map<int, std::vector<int>>& adjList) {
    for (const auto& entry : adjList) {
        std::cout << entry.first << " -> ";
        for (int neighbor : entry.second) {
            std::cout << neighbor << " ";
        }
        std::cout << std::endl;
    }
}

int main() {
    int n = 8;
    std::vector<std::pair<int, int>> edges = {{0, 1}, {1, 2}, {0, 3}, {3, 4}, {3, 6}, {3, 7}, {4, 2}, {4, 5}, {5, 2}};

    // Convert Array of Edges -> Adjacency Matrix
    auto adjacencyMatrix = buildAdjacencyMatrix(n, edges);
    std::cout << "Adjacency Matrix:" << std::endl;
    printMatrix(adjacencyMatrix);

    // Convert Array of Edges -> Adjacency List
    auto adjacencyList = buildAdjacencyList(edges);
    std::cout << "Adjacency List:" << std::endl;
    printAdjacencyList(adjacencyList);

    // Depth First Search (DFS) - Recursive
    std::cout << "DFS Recursive:" << std::endl;
    std::unordered_set<int> seen;
    dfsRecursive(0, adjacencyList, seen);
    std::cout << std::endl;

    // Depth First Search (DFS) - Iterative
    std::cout << "DFS Iterative:" << std::endl;
    dfsIterative(0, adjacencyList);
    std::cout << std::endl;

    // Breadth First Search (BFS)
    std::cout << "BFS:" << std::endl;
    bfs(0, adjacencyList);
    std::cout << std::endl;

    return 0;
}

File: Python/11 - Graphs - Edge List, Adjacency Matrix, Adjacency List, DFS, BFS - Greg Hogg DSA Course Materials Lecture 11.py
# Graphs - Edge List, Adjacency Matrix, Adjacency List, DFS, BFS - Greg Hogg DSA Course Materials Lecture 11

# The Colab Notebook can be viewed here: https://colab.research.google.com/drive/1V5Afu8zDQLJpPXQLMyocMm93GB2qdOJV

# Array of Edges (Directed) [Start, End]
n = 8
A = [[0, 1], [1, 2], [0, 3], [3, 4], [3, 6], [3, 7], [4, 2], [4, 5], [5, 2]]

print(A)

# Convert Array of Edges -> Adjacency Matrix

M = []
for i in range(n):
  M.append([0] * n)

for u, v in A:
  M[u][v] = 1

  # Uncomment the following line if the graph is undirected
  # M[v][u] = 1

print(M)

# Convert Array of Edges -> Adjacency List
from collections import defaultdict

D = defaultdict(list)

for u, v in A:
  D[u].append(v)
  # Uncomment the following line if the graph is undirected
  # D[v].append(u)

print(D)

print(D[3])

# DFS with Recursion - O(V + E) where V is the number of nodes and E is the number of edges

def dfs_recursive(node):
  print(node)
  for nei_node in D[node]:
    if nei_node not in seen:
      seen.add(nei_node)
      dfs_recursive(nei_node)

source = 0
seen = set()
seen.add(source)
dfs_recursive(source)

# Iterative DFS with Stack - O(V + E)

source = 0

seen = set()
seen.add(source)
stack = [source]

while stack:
  node = stack.pop()
  print(node)
  for nei_node in D[node]:
    if nei_node not in seen:
      seen.add(nei_node)
      stack.append(nei_node)

# BFS (Queue) - O(V + E)

source = 0

from collections import deque

seen = set()
seen.add(source)
q = deque()
q.append(source)

while q:
  node = q.popleft()
  print(node)
  for nei_node in D[node]:
    if nei_node not in seen:
      seen.add(nei_node)
      q.append(nei_node)

class Node:
  def __init__(self, value):
    self.value = value
    self.neighbors = []

  def __str__(self):
    return f'Node({self.value})'

  def display(self):
    connections = [node.value for node in self.neighbors]
    return f'{self.value} is connected to: {connections}'

A = Node('A')
B = Node('B')
C = Node('C')
D = Node('D')

A.neighbors.append(B)
B.neighbors.append(A)

C.neighbors.append(D)
D.neighbors.append(C)

print(B.display())
